httpd/http_server.c

```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "daemon/common.h"
+#include "http_server.h"
+#include "h2o.h"
+
+#include "h2o_utils.h"
+
+static h2o_globalconf_t config;
+static h2o_context_t ctx;
+static h2o_accept_ctx_t accept_ctx;
+
+#define CONTENT_JSON_UTF8 H2O_STRLIT("application/json; charset=utf-8")
+#define CONTENT_TEXT_UTF8 H2O_STRLIT("text/plain; charset=utf-8")
+#define NBUF_INITIAL_SIZE_RESP (4096)
+#define API_V1_PREFIX "/api/v1/"
+#define HOST_SELECT_PREFIX "/host/"
+
+#define HTTPD_CONFIG_SECTION "httpd"
+#define HTTPD_ENABLED_DEFAULT false
+
+static void on_accept(uv_stream_t *listener, int status)
+{
+    uv_tcp_t *conn;
+    h2o_socket_t *sock;
+
+    if (status != 0)
+        return;
+
+    conn = h2o_mem_alloc(sizeof(*conn));
+    uv_tcp_init(listener->loop, conn);
+
+    if (uv_accept(listener, (uv_stream_t *)conn) != 0) {
+        uv_close((uv_handle_t *)conn, (uv_close_cb)free);
+        return;
+    }
+
+    sock = h2o_uv_socket_create((uv_stream_t *)conn, (uv_close_cb)free);
+    h2o_accept(&accept_ctx, sock);
+}
+
+static int create_listener(const char *ip, int port)
+{
+    static uv_tcp_t listener;
+    struct sockaddr_in addr;
+    int r;
+
+    uv_tcp_init(ctx.loop, &listener);
+    uv_ip4_addr(ip, port, &addr);
+    if ((r = uv_tcp_bind(&listener, (struct sockaddr *)&addr, 0)) != 0) {
+        fprintf(stderr, "uv_tcp_bind:%s\n", uv_strerror(r));
+        goto Error;
+    }
+    if ((r = uv_listen((uv_stream_t *)&listener, 128, on_accept)) != 0) {
+        fprintf(stderr, "uv_listen:%s\n", uv_strerror(r));
+        goto Error;
+    }
+
+    return 0;
+Error:
+    uv_close((uv_handle_t *)&listener, NULL);
+    return r;
+}
+
+static int ssl_init()
+{
+    if (!config_get_boolean(HTTPD_CONFIG_SECTION, "ssl", false))
+        return 0;
+
+    char default_fn[FILENAME_MAX + 1];
+
+    snprintfz(default_fn,  FILENAME_MAX, "%s/ssl/key.pem",  netdata_configured_user_config_dir);
+    const char *key_fn  = config_get(HTTPD_CONFIG_SECTION, "ssl key", default_fn);
+
+    snprintfz(default_fn, FILENAME_MAX, "%s/ssl/cert.pem", netdata_configured_user_config_dir);
+    const char *cert_fn = config_get(HTTPD_CONFIG_SECTION, "ssl certificate",  default_fn);
+
+    accept_ctx.ssl_ctx = SSL_CTX_new(SSLv23_server_method());
```

Comment:
A suggestion to avoid issues with OpenSSL version. As you can see in our current [code](https://github.com/netdata/netdata/blob/513452aa447759acecb16504673db7dce40ffaa2/libnetdata/socket/security.c#L121-L125), we are initializing contexts according OpenSSL version, because this function used here can be deprecated in a near future, please, follow the same direction for all SSL functions we are using in this new file.

Reply:
It doesn't really matter because new OpenSSL handles it like this:
```
#define SSLv23_method           TLS_method
#define SSLv23_server_method    TLS_server_method
#define SSLv23_client_method    TLS_client_method
```

so if you use `SSLv23_server_method` old SSL will really use `SSLv23_server_method` and new one will translate it to `TLS_server_method`  but I added the #ifdef you suggested to make it explicit

Reply:
All right, I missed that.